File: backend/src/domain/antiLinkSanitizer.js
/**
 * Filtro de seguridad anti-desintermediación para perfiles y mensajes.
 * Detecta y bloquea URLs, correos electrónicos y números de teléfono.
 */
export class AntiLinkSanitizer {
  // Regex sin flag global para pruebas deterministas
  static get URL_REGEX() {
    return /(https?:\/\/[^\s]+|www\.[^\s]+|\b[a-zA-Z0-9.-]+\.(?:com|bo|org|net|edu|io|co|me|dev)\b)/i;
  }

  static get EMAIL_REGEX() {
    return /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/i;
  }

  static get PHONE_REGEX() {
    return /(\+?\d{1,3}[\s-]?)?\(?\d{2,3}\)?[\s-]?\d{4,5}[\s-]?\d{3,5}|\b\d{7,10}\b/;
  }

  static validate(text) {
    if (!text) return true;

    // Evaluamos primero emails para mensajes específicos
    if (this.EMAIL_REGEX.test(text)) {
      throw new Error("Violación de política: No está permitido incluir correos electrónicos.");
    }
    if (this.URL_REGEX.test(text)) {
      throw new Error("Violación de política: No está permitido incluir enlaces externos (URLs).");
    }
    if (this.PHONE_REGEX.test(text)) {
      throw new Error("Violación de política: No está permitido compartir números telefónicos.");
    }

    return true;
  }

  static sanitize(text) {
    if (!text) return "";

    const urlGlobal = new RegExp(this.URL_REGEX.source, "gi");
    const emailGlobal = new RegExp(this.EMAIL_REGEX.source, "gi");
    const phoneGlobal = new RegExp(this.PHONE_REGEX.source, "g");

    return text
      .replace(emailGlobal, "[DATO DE CONTACTO BLOQUEADO]")
      .replace(urlGlobal, "[DATO DE CONTACTO BLOQUEADO]")
      .replace(phoneGlobal, "[DATO DE CONTACTO BLOQUEADO]");
  }
}

File: src/core/services/anti_link_sanitizer.py
import re
from src.core.domain.exceptions import DisintermediationViolationException


class AntiLinkSanitizerService:
    """
    Filtro de seguridad contra la desintermediación.
    Detecta, valida y censura URLs, correos electrónicos y números telefónicos.
    """

    URL_PATTERN = re.compile(r"(https?://\S+|www\.\S+|\b[a-zA-Z0-9.-]+\.(?:com|bo|org|net|edu|io|co|me|dev)\b)", re.IGNORECASE)
    EMAIL_PATTERN = re.compile(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b")
    PHONE_PATTERN = re.compile(r"(\+?\d{1,3}[\s-]?)?\(?\d{2,3}\)?[\s-]?\d{4,5}[\s-]?\d{3,5}|\b\d{7,10}\b")

    def validate(self, text: str) -> bool:
        """
        Valida que el texto no contenga enlaces externos ni datos de contacto.
        Lanza DisintermediationViolationException si se detecta infracción.
        """
        if not text:
            return True

        if self.URL_PATTERN.search(text):
            raise DisintermediationViolationException("Se ha detectado un enlace web externo. No está permitido incluir URLs.")

        if self.EMAIL_PATTERN.search(text):
            raise DisintermediationViolationException("Se ha detectado un correo electrónico. La comunicación debe ser interna.")

        if self.PHONE_PATTERN.search(text):
            raise DisintermediationViolationException("Se ha detectado un número telefónico. La comunicación debe ser interna.")

        return True

    def sanitize(self, text: str) -> str:
        """Reemplaza cualquier dato de contacto detectado por una advertencia de bloqueo."""
        if not text:
            return ""

        result = self.URL_PATTERN.sub("[DATO DE CONTACTO BLOQUEADO]", text)
        result = self.EMAIL_PATTERN.sub("[DATO DE CONTACTO BLOQUEADO]", result)
        result = self.PHONE_PATTERN.sub("[DATO DE CONTACTO BLOQUEADO]", result)
        return result
